### DISCUSSION TITLE
Accessing last data entries

### DISCUSSION THREAD

**Discussion by maxblomberg** [jspsych/jsPsych#965] [Q&A] [answered]
Hello everybody,

I am trying to create an experiment where the trial duration (or inter-trial-interval ITI) depends on whether the answer in the previous trial was correct. I am using the audio-button-response plugin and have hosted the experiment at [cognition.run ](https://yu2zf7c4di.cognition.run/)
Unfortunately I have a problem accessing the data entries of previous trials. I have tried different forms of jsPsych.data.get(), but nothing worked so far for me. It seems like accessing the previous data entries is just ignored and the respective variables remain undefined.

Does anyone else have an idea?

Thank you so much in advance!

```javascript
var audio = ["d1.mp3", "d2.mp3", "d3.mp3", "d4.mp3", "d5.mp3",
"d6.mp3", "d7.mp3", "d8.mp3", "d9.mp3", "d10.mp3", "d11.mp3", "d12.mp3",
"d13.mp3", "d14.mp3", "d15.mp3", "d16.mp3", "d17.mp3", "d18.mp3"];

var images = ["one.PNG", "two.PNG", "three.PNG", "four.PNG", "five.PNG",
"six.PNG", "seven.PNG", "eight.PNG", "nine.PNG", "ten.PNG", "eleven.PNG", "twelve.PNG",
"thirteen.PNG", "fourteen.PNG", "fifteen.PNG", "sixteen.PNG", "seventeen.PNG", "eighteen.PNG"];

var audio_dur = 816; // duration of audio files
var maxITI = 3000;
var startITI = 2000;
var minITI = 1000;
var ITIadaptationCriterium = 4;
var ITIchange = 100;


var test_stimuli = [
  {stimulus: "d1.mp3", data: {correct_response: "0"}}, 
  {stimulus: "d2.mp3", data: {correct_response: "1"}}, 
  {stimulus: "d3.mp3", data: {correct_response: "2"}},  
  {stimulus: "d4.mp3", data: {correct_response: "3"}},  
  {stimulus: "d5.mp3", data: {correct_response: "4"}}, 
  {stimulus: "d6.mp3", data: {correct_response: "5"}},  
  {stimulus: "d7.mp3", data: {correct_response: "6"}},  
  {stimulus: "d8.mp3", data: {correct_response: "7"}}, 
  {stimulus: "d9.mp3", data: {correct_response: "8"}},  
  {stimulus: "d10.mp3", data: {correct_response: "9"}},  
  {stimulus: "d11.mp3", data: {correct_response: "10"}},  
  {stimulus: "d12.mp3", data: {correct_response: "11"}}, 
  {stimulus: "d13.mp3", data: {correct_response: "12"}},  
  {stimulus: "d14.mp3", data: {correct_response: "13"}},  
  {stimulus: "d15.mp3", data: {correct_response: "14"}},  
  {stimulus: "d16.mp3", data: {correct_response: "15"}}, 
  {stimulus: "d17.mp3", data: {correct_response: "16"}}, 
  {stimulus: "d18.mp3", data: {correct_response: "17"}}
 ];


/* interaction check */
var interaction = {
    type: 'html-button-response',
    stimulus: '<p>Zum Beginnen klicken Sie bitte auf "Start"</p>',
    choices: ['Start'],
};

var test = {
      type: 'audio-button-response',
      stimulus: jsPsych.timelineVariable("stimulus"),
      choices: images,
      button_html: '<img src="%choice%" style="max-width:70%;" />',
      response_ends_trial: false,
      data: jsPsych.timelineVariable("data"),
      trial_duration: function(){
        if (jsPsych.data.get().count() > 2){ 
          trial_duration = jsPsych.data.get().last(1).values()[0].nextITI  + audio_dur; // modify trial duration based on adjusted ITI
        } else {
          trial_duration = startITI + audio_dur; // trial duration set for first 2 trials
        }
        return trial_duration
      },
      on_finish: function(data){
            
        data.correct = data.button_pressed == data.correct_response; // was response correct?
        data.currITI = trial_duration - audio_dur; // store current ITI
            
          if (jsPsych.data.get().count() > 2){
              
              // check for response and adjust counters resp.
              if (data.correct == true){ 
		      	  data.consecutiveErrors = 0;
			        data.count_consecutiveErrors = 0;
			        data.count_consecutiveCorrect = jsPsych.data.get().last(1).values()[0].count_consecutiveCorrect + 1; // add 1 for correct response
			        data.consecutiveCorrect = jsPsych.data.get().last(1).values()[0].consecutiveCorrect + 1;
		          } else {
			        data.consecutiveCorrect = 0;
			        data.count_consecutiveCorrect = 0;
			        data.count_consecutiveErrors = jsPsych.data.get().last(1).values()[0].count_consecutiveErrors + 1; // add 1 for incorrect response
			        data.consecutiveErrors = jsPsych.data.get().last(1).values()[0].consecutiveErrors + 1;
		          }
		          
		          // check if ITI needs to be adjusted
		          if (data.count_consecutiveCorrect >= ITIadaptationCriterium){
			        data.nextITI = jsPsych.data.get().last(1).values()[0].nextITI - ITIchange; //decrease the ITI by 100ms (default) for 4 (default) consecutive correct responses
			          if (data.nextITI < minITI){ //check that next ITI is not lower than the minimum ITI
				        data.nextITI = minITI;
			          }
			        data.count_consecutiveCorrect = 0; //reset the counter for consecutive correct responses
		          } else if (data.count_consecutiveErrors >= ITIadaptationCriterium){
        			data.nextITI = jsPsych.data.get().last(1).values()[0].nextITI + ITIchange; //increase the ITI by 100ms (default) for 4 (default) consecutive error responses
          			if (data.nextITI > maxITI){
          			data.nextITI = maxITI; //check that next ITI does not exceed the maximum allowed ITI
          			}
        			data.count_consecutiveErrors = 0; //reset the counter for consecutive error responses
		          } else {
		          data.nextITI = jsPsych.data.get().last(1).values()[0].nextITI; // if ITI is not adjusted, just use the previous one
		          }
		          
          } else { // for the first two trials define static values
          data.consecutiveCorrect = 0;
	  data.count_consecutiveCorrect = 0;
	  data.count_consecutiveErrors = 0;
	  data.consecutiveErrors = 0;
	  data.nextITI = startITI;
			    
          }            
      },
    };
      
      
var test_procedure = {
    timeline: [test],
    timeline_variables: test_stimuli,
    randomize_order: true,
    repetitions: 1
    };


    /* start the experiment */
jsPsych.init({
  preload_audio: audio,
  preload_images: images,
  timeline: [interaction, test_procedure],
  use_webaudio: true,
  on_finish: function() {
      jsPsych.data.displayData();
  }
})
```

**Comment by vijaymarupudi** [comment]
FYI, you are creating a global variable here, that usually leads to very confusing bugs. 

```javascript
      trial_duration: function(){
        if (jsPsych.data.get().count() > 2){ 
          trial_duration = jsPsych.data.get().last(1).values()[0].nextITI  + audio_dur; // modify trial duration based on adjusted ITI
        } else {
          trial_duration = startITI + audio_dur; // trial duration set for first 2 trials
        }
        return trial_duration
      },
```

You use the global variable in the `on_finish` callback, I would instead retrieve the `trial_duration` from the data variable passed to the callback.

Anyways, I can help you debug it! Nothing seems obviously wrong to me currently. Could you provide the stack trace and error message, and/or a working copy of the experiment?

**Reply by maxblomberg** [reply]
Thanks for your reply! Attached is the console error when running the experiment locally (I don't get a console log when running it on cognition.run). The experiment works when setting `trial_duration` to a constant value, e.g. `trial_duration = startITI + audio_dur`. However, the dynamic data variables, e.g. `data.count_consecutiveCorrect` end up with `null`, which lead me to the idea that I might somehow access those values in the wrong way. 
I don't know how to provide the stack trace - can you tell me how to do that if the error message shouldn't be enough?

![console error](https://user-images.githubusercontent.com/69300021/89921294-cc5bd280-dbfd-11ea-8d84-84768fc057ec.PNG)

**Reply by vijaymarupudi** [reply]
Could you send a working copy of the experiment. I suspect the problem is with deal with the asynchronity of javascript and race conditions, both of which are hard to debug without a working copy! Those errors seem very scary haha

**Reply by maxblomberg** [reply]
Ah yes, sorry! Attached is the zip-file containing the working experiment with audio/image files which should work when the js-psych-folder is in the same folder. The idea of the experiment is to listen to a number and then click on the respective button (sorry audio of numbers is in German).

[pasat_sham_training.zip](https://github.com/jspsych/jsPsych/files/5059069/pasat_sham_training.zip)

**Comment by vijaymarupudi** [comment]
Don't apologize for that  :)

Umm, so I think the issue is that nextITI is undefined due to your logic, not because of jsPsych. I tried to parse through the complicated logic you have going on, but it's a little difficult because I do not know the exact paradigm.

What you have to do is ensure that after every trial, at the end of the `on_finish` function, `data.nextITI` is defined. Currently, there are logical loopholes, i.e. there is a way where `data.nextITI` is not defined by the end of the `on_finish` function. I would set a `console.log(data)` at the end of the function and try stuff out.

Sorry I couldn't be of much help! A word of advice, I would refactor similar code into a variable, like setting `jsPsych.data.get().last(1).values()[0]` to a variable called `lastTrial` and then using that.  That makes it easier to debug.

**Comment by maxblomberg** [comment, marked as answer]
Solution found: When using the `on_finish` callback-function, jsPsych already treats the current trial to have finished, e.g. `jsPsych.data.get().last(1).values()[0]` would refer to the same trial, in which the variables in question are yet not defined. Just had to change that part in the `on_finish` block to `jsPsych.data.get().last(2).values()[0]`.

Anyways, thanks for your help Vijay!

**Reply by vijaymarupudi** [reply]
Ah I see! Good to know you've solved it!!

~ Vijay
